integration-testing/cypress/e2e/jupyter.cy.js

```diff
@@ -0,0 +1,55 @@
+describe('Publishing Jupyter Notebook', () => {
+  beforeEach(() => {
+  });
+
+// all tests had to be run in a single 
+  it('Publish button loads', () => {
+    cy.visit('http://client:9999/tree/integration-testing/content/notebook/stock-report-jupyter.ipynb');
+    cy.get('button[data-jupyter-action="rsconnect_jupyter:publish"]').click();
+    cy.get('a[id="publish-to-connect"]').should('be.visible')
+  });
+  // wait is required after every action, cypress is too fast for jupyter
+  // https://github.com/cypress-io/cypress/issues/249
+  it('Add Server', () => {
+    cy.visit('http://client:9999/tree/integration-testing/content/notebook/stock-report-jupyter.ipynb');
+    cy.wait(1000);
+    cy.get('a[id="publish-to-connect"]').click({ force: true });
+    cy.wait(1000);
+    cy.get('input[id="rsc-server"]').clear().type('http://connect:3939');
+    cy.get('input[id="rsc-api-key"]').clear().type(Cypress.env('api_key'));
+    cy.get('input[id="rsc-servername"]').clear().type('localhost');
+    cy.get('a[class="btn btn-primary"]').contains(' Add Server').click();
+    cy.wait(1000);
+    cy.get('span[class="help-block"]').should('not.have.text',"Unable to verify");
+  });
+  it('Publish Content', () => {
+    cy.visit('http://client:9999/tree/integration-testing/content/notebook/stock-report-jupyter.ipynb');
+    cy.wait(1000);
+    cy.get('a[id="publish-to-connect"]').click({ force: true });
+    cy.wait(1000);
+    cy.get('button[id="rsc-add-files"]').click();
+    cy.wait(1000);
+    cy.get('input[name="quandl-wiki-tsla.json.gz"]').click();
+    cy.wait(1000);
+    cy.get('button[id="add-files-dialog-accept"]').click();
+    cy.wait(1000);
+    cy.get('li[class="list-group-item"]').first().should('have.text'," quandl-wiki-tsla.json.gz");
+    cy.wait(1000);
+    cy.get('a[class="btn btn-primary"]').last().should('have.text',"Publish").click({ force: true });
+    cy.wait(1000);
+    cy.get('input[name="location"]').first().click();
+    cy.wait(1000);
+    cy.get('a[class="btn btn-primary"]').last().should('have.text',"Next").click();
+    cy.wait(1000);
+    cy.get('a[class="btn btn-primary"]').last().should('have.text',"Publish").click();
+    cy.wait(1000);
```

Comment:
Are the `cy.wait`'s required? I thought that `cy.get` waits until elements appear.

Reply:
Is it possible to wait for some action to complete, instead of an arbitrary time? For example, if the click() action triggers a network request.

Reply:
I've tried waiting for different elements to load but cypress is simply running too fast, the only way to avoid flakiness here I think is to hardcode `cy.wait()` each time the modal loads. 

Here is a screenshot of what's happening.
![Screen Shot 2023-02-09 at 3 23 09 PM](https://user-images.githubusercontent.com/41300090/217935001-389e8e97-8621-4742-84a3-45c3a572457e.png)

Cypress code (after removing `cy.wait()`):
```
    cy.get('button[id="rsc-add-files"]')
      .should('be.visible')
      .and('contain',"Select Files...")
      .click();
```
This says the Select Files button should be visible and contain "Select Files..." and if so, click it.  But Cypress "clicks" it before the element is ready and then can't find the files from the _next_ page and fails there.

This seems to be Cypress' biggest issue, there are loads of forum posts and blogs about it and most come to the conclusion, just add a wait(). 

I rather like this blog post, it's right in the title, `waiting-in-cypress-and-how-to-avoid-it` but then immediately admits, it's probably best to add the wait().
https://filiphric.com/waiting-in-cypress-and-how-to-avoid-it